daemon/delete.go

```diff
@@ -129,6 +129,9 @@ func (daemon *Daemon) commonRm(container *Container, forceRemove bool) (err erro
 		if err != nil && forceRemove {
 			daemon.idIndex.Delete(container.ID)
 			daemon.containers.Delete(container.ID)
+			if _, err := os.Stat(container.root); err != nil {
```

Comment:
This is in a defer that for container rm.
If you `docker rm -f` a container and there is an error, this tells docker to de-ref the container so it's not listed anymore... except we didn't remove the actual on disk config so when the daemon restarts the container is back.